File: game/input/Input.ts
// 输入系统：键盘(WASD+方向键)/手柄左摇杆/虚拟摇杆 合成移动方向；Esc/Start 暂停触发；攻击按住状态
import { KeyName, Keyboard, ButtonName, AxisName, Controller, Vec2 } from 'Dora';
import type { VirtualJoystick } from 'game/ui/VirtualJoystick';

export class InputSystem {
	private joystick: VirtualJoystick | undefined = undefined;
	private _moveDir: Vec2.Type = Vec2(0, 0);
	private _pausePressed = false;
	private _attackHeld = false;

	// 设置虚拟摇杆引用（主控在创建后调用）
	attachJoystick(j: VirtualJoystick): void {
		this.joystick = j;
	}

	get moveDir(): Vec2.Type {
		return this._moveDir;
	}

	get pausePressed(): boolean {
		return this._pausePressed;
	}

	get isAttackHeld(): boolean {
		return this._attackHeld;
	}

	// 每帧轮询（主控 update 循环调用）
	update(): void {
		this._pausePressed = false;

		let dx = 0;
		let dy = 0;

		// 键盘 WASD + 方向键（isKeyPressed 为按住状态）
		if (Keyboard.isKeyPressed(KeyName.W) || Keyboard.isKeyPressed(KeyName.Up)) dy += 1;
		if (Keyboard.isKeyPressed(KeyName.S) || Keyboard.isKeyPressed(KeyName.Down)) dy -= 1;
		if (Keyboard.isKeyPressed(KeyName.A) || Keyboard.isKeyPressed(KeyName.Left)) dx -= 1;
		if (Keyboard.isKeyPressed(KeyName.D) || Keyboard.isKeyPressed(KeyName.Right)) dx += 1;

		// 手柄左摇杆
		const ax = Controller.getAxis(0, AxisName.LeftX);
		const ay = Controller.getAxis(0, AxisName.LeftY);
		if (Math.abs(ax) > 0.15 || Math.abs(ay) > 0.15) {
			dx += ax;
			dy += ay;
		}

		// 虚拟摇杆（移动端触控）
		if (this.joystick !== undefined && this.joystick.isVisible) {
			const jv = this.joystick.moveDir;
			dx += jv.x;
			dy += jv.y;
		}

		// 归一化（保留手柄/摇杆的幅度缩放）
		const len = Math.sqrt(dx * dx + dy * dy);
		if (len > 1) {
			dx /= len;
			dy /= len;
		}
		this._moveDir = Vec2(dx, dy);

		// 暂停：Esc 或手柄 Start 一帧触发
		if (Keyboard.isKeyDown(KeyName.Escape) || Controller.isButtonDown(0, ButtonName.Start)) {
			this._pausePressed = true;
		}

		// 攻击按住：空格/手柄 X（占位，攻击时机由战斗域决定）
		this._attackHeld =
			Keyboard.isKeyPressed(KeyName.Space) || Controller.isButtonDown(0, ButtonName.X);
	}
}

// 单例导出：主控创建后调用 inputSystem.update()
export const inputSystem = new InputSystem();

File: game/input/Input.lua
-- [ts]: Input.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__Class = ____lualib.__TS__Class -- 1
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor -- 1
local __TS__New = ____lualib.__TS__New -- 1
local ____exports = {} -- 1
local ____Dora = require("Dora") -- 2
local Keyboard = ____Dora.Keyboard -- 2
local Controller = ____Dora.Controller -- 2
local Vec2 = ____Dora.Vec2 -- 2
____exports.InputSystem = __TS__Class() -- 5
local InputSystem = ____exports.InputSystem -- 5
InputSystem.name = "InputSystem" -- 5
function InputSystem.prototype.____constructor(self) -- 5
	self.joystick = nil -- 6
	self._moveDir = Vec2(0, 0) -- 7
	self._pausePressed = false -- 8
	self._attackHeld = false -- 9
end -- 5
function InputSystem.prototype.attachJoystick(self, j) -- 12
	self.joystick = j -- 13
end -- 12
function InputSystem.prototype.update(self) -- 29
	self._pausePressed = false -- 30
	local dx = 0 -- 32
	local dy = 0 -- 33
	if Keyboard:isKeyPressed("W") or Keyboard:isKeyPressed("Up") then -- 33
		dy = dy + 1 -- 36
	end -- 36
	if Keyboard:isKeyPressed("S") or Keyboard:isKeyPressed("Down") then -- 36
		dy = dy - 1 -- 37
	end -- 37
	if Keyboard:isKeyPressed("A") or Keyboard:isKeyPressed("Left") then -- 37
		dx = dx - 1 -- 38
	end -- 38
	if Keyboard:isKeyPressed("D") or Keyboard:isKeyPressed("Right") then -- 38
		dx = dx + 1 -- 39
	end -- 39
	local ax = Controller:getAxis(0, "leftx") -- 42
	local ay = Controller:getAxis(0, "lefty") -- 43
	if math.abs(ax) > 0.15 or math.abs(ay) > 0.15 then -- 43
		dx = dx + ax -- 45
		dy = dy + ay -- 46
	end -- 46
	if self.joystick ~= nil and self.joystick.isVisible then -- 46
		local jv = self.joystick.moveDir -- 51
		dx = dx + jv.x -- 52
		dy = dy + jv.y -- 53
	end -- 53
	local len = math.sqrt(dx * dx + dy * dy) -- 57
	if len > 1 then -- 57
		dx = dx / len -- 59
		dy = dy / len -- 60
	end -- 60
	self._moveDir = Vec2(dx, dy) -- 62
	if Keyboard:isKeyDown("Escape") or Controller:isButtonDown(0, "start") then -- 62
		self._pausePressed = true -- 66
	end -- 66
	self._attackHeld = Keyboard:isKeyPressed("Space") or Controller:isButtonDown(0, "x") -- 70
end -- 29
__TS__SetDescriptor( -- 29
	InputSystem.prototype, -- 29
	"moveDir", -- 29
	{get = function(self) -- 29
		return self._moveDir -- 17
	end}, -- 17
	true -- 17
) -- 17
__TS__SetDescriptor( -- 17
	InputSystem.prototype, -- 17
	"pausePressed", -- 17
	{get = function(self) -- 17
		return self._pausePressed -- 21
	end}, -- 21
	true -- 21
) -- 21
__TS__SetDescriptor( -- 21
	InputSystem.prototype, -- 21
	"isAttackHeld", -- 21
	{get = function(self) -- 21
		return self._attackHeld -- 25
	end}, -- 25
	true -- 25
) -- 25
____exports.inputSystem = __TS__New(____exports.InputSystem) -- 76
return ____exports -- 76